gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/TraversalHelper.java

```diff
@@ -465,7 +465,11 @@ public static boolean anyStepRecursively(final Predicate<Step> predicate, final
      */
     public static void applyTraversalRecursively(final Consumer<Traversal.Admin<?, ?>> consumer, final Traversal.Admin<?, ?> traversal) {
         consumer.accept(traversal);
-        for (final Step<?, ?> step : traversal.getSteps()) {
+
+        // we get accused of concurrentmodification if we try a for(Iterable)
```

Comment:
I think it can be simplified into:

```java
traversal.getSteps().stream().filter(step -> step instanceof TraversalParent).forEach(step -> {
    for (final Traversal.Admin<?, ?> local : ((TraversalParent) step).getLocalChildren()) {
        applyTraversalRecursively(consumer, local);
    }
    for (final Traversal.Admin<?, ?> global : ((TraversalParent) step).getGlobalChildren()) {
        applyTraversalRecursively(consumer, global);
    }
});
```

I'm not sure I understand why you were getting concurrent modification exception.

Reply:
I guess you get concurrent modification because strategies may modify the list of steps while the iteration is happening...

Reply:
ah, I see...